staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+            return;
+        }
+        self.last_epoch_height = epoch_height;
+
+        let total_balance =
+            env::account_locked_balance() + env::account_balance() - env::attached_deposit();
+        let last_total_balance = self.last_account_balance + self.last_locked_account_balance;
+
+        let total_reward = total_balance.saturating_sub(last_total_balance);
+        // The total stake is increased by both staked and unstaked rewards.
+        self.total_staked_balance += total_reward;
+
+        self.last_locked_account_balance = env::account_locked_balance();
+        self.last_account_balance = env::account_balance();
+    }
+
+    /// Call to deposit money.
+    #[payable]
+    pub fn deposit(&mut self) {
+        self.ping();
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+        account.unstaked += env::attached_deposit();
+        self.save_account(&account_id, account);
+
+        if self.total_staked_balance > self.last_locked_account_balance {
+            self.restake();
+        }
+    }
+
+    /// Withdraws the non staked balance for given account.
+    pub fn withdraw(&mut self, amount: U128) {
+        self.ping();
+
+        let amount = amount.into();
+        assert!(amount > 0, "Withdrawal amount should be positive");
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+        assert!(
+            account.unstaked >= amount,
+            "Not enough unstaked balance to withdraw"
+        );
+        assert!(
+            account.unstaked_available_epoch_height <= env::epoch_height(),
+            "The unstaked balance is not yet available due to unstaking delay"
+        );
+        account.unstaked -= amount;
+        self.save_account(&account_id, account);
+        Promise::new(account_id).transfer(amount);
+
+        if self.total_staked_balance > self.last_locked_account_balance {
+            self.restake();
+        }
+    }
+
+    /// Stakes previously deposited money by given account on this account.
+    pub fn stake(&mut self, amount: U128) {
+        self.ping();
+
+        let amount: Balance = amount.into();
+        assert!(amount > 0, "Staking amount should be positive");
+
+        let account_id = env::predecessor_account_id();
+        let mut account = self.get_account(&account_id);
+
+        let num_shares = if self.total_staked_balance > 0 {
+            // price = total_staked / total_shares
+            // Price is fixed
+            // (total_staked + amount) / (total_shares + num_shares) = total_staked / total_shares
+            // (total_staked + amount) * total_shares = total_staked * (total_shares + num_shares)
+            // amount * total_shares = total_staked * num_shares
+            // num_shares = amount * total_shares / total_staked
+            // NOTE: Number of shares the account gets is rounded up, but the gas rebate within
```

Comment:
hm, that's a bit dangerous IMO. what if governance removes rebates? why not round down?

Reply:
It would mean the lockup contract might lose some tokens during stake/unstake call, which might break staking logic. Within lockup contract, we should either get the amount of tokens that we staked or not rely store internal values. Then there are no issue of losing some small amount of tokens during staking.

Reply:
Obsolete